### ISSUE TITLE
devicetree: need save/restore support for devicetree data

### ISSUE THREAD

**Issue by pabigot** [zephyrproject-rtos/zephyr#26575]
tl;dr: I need the EDT object that's constructed during devicetree processing to be persisted to disk so I can read it back and use its data to generate optimized dependency information to be used in the final application link.

The parts of #26388 related to using devicetree and macros to add dependency information to the device structure work well.  However the actual values of the device handles are not well chosen: devicetree ordinals are captured without regard to what devices are actually used in the application.  Thus there are gaps in the handle space, and dependencies are recorded to nodes that have no corresponding device.

The handle assignment can be cleaned up with a post-process step on `zephyr_prebuilt.elf`, much like what's done for kernel objects and system call support.  The concept is to link that image with the dependency arrays filled in from sections that encode the original ordinals, then to read the arrays from that image and map the ordinal-based handles to handles based on the devices used.  This would also mean that the handle for a specific device could be its offset from the start of the device table, making it trivial to map from the handle to the device pointer.

To accomplish this we'd need to construct the subgraph of the devicetree dependency graph that contains only the root and nodes that are present in the application.  To do that we need the devicetree state that includes the original dependency graph (as well as all the other data, for reasons).

To do that we need a way to save that state to a file in the build area, then load an EDT instance in post-processing.
